torch/distributed/pipelining/stage.py

```diff
@@ -489,21 +515,41 @@ def backward_maybe_with_nosync(self, bwd_kwargs: Dict):
                         )
                     )
                 )
-                grads_input = stage_backward(**bwd_kwargs)
+                result = perform_backward(backward_type)()
             else:
                 with self.submod.no_sync():  # type: ignore[operator]
-                    grads_input = stage_backward(**bwd_kwargs)
+                    result = perform_backward(backward_type)()
         # If submod is a FSDP module
         elif isinstance(self.submod, FSDPModule):
-            self.submod.set_is_last_backward(last_backward)
-            self.submod.set_requires_gradient_sync(last_backward)
-            grads_input = stage_backward(**bwd_kwargs)
+            self.submod.set_is_last_backward(False)
+            self.submod.set_reshard_after_backward(False)
+            self.submod.set_requires_gradient_sync(False)
+            result = perform_backward(backward_type)()
+            if last_backward:
+                # Manually call post backward for FSDP
+                def run_post_backward(fsdp_module: FSDPModule) -> None:
+                    fsdp_module.set_is_last_backward(True)
+                    fsdp_module.set_reshard_after_backward(True)
+                    fsdp_module.set_requires_gradient_sync(True)
+                    fsdp_state = fully_shard.state(fsdp_module)
+                    for state in fsdp_state._state_ctx.all_states:
+                        if state._fsdp_param_group:
+                            state._fsdp_param_group.post_backward()
```

Comment:
is this advocating a new fsdp api, like `model.reduce_grad` ?

Reply:
I guess this new api could be useful if users call `autograd.grad()` instead of `autograd.backward()` and still want to reduce the gradients like what I am doing.

In practice, I'm not sure how needed this is for FSDP users, maybe my case is just a pretty unconventional use case.

Reply:
I am asking `._state_ctx` and `._fsdp_param_group` are private. there is some risk of future change. but it's not a immeidate blocker for this PR

Reply:
It's probably better if we use a private api and rely on ci composability tests to keep it working rather than make a new public api that we don't really need

Reply:
Agree. Would be nice if we don't have to get into FSDP details.